backend/tests/Feature/PublicationTest.php

```diff
@@ -178,388 +171,151 @@ public function testQueryFiltersHiddenItems()
         $this->assertCount(2, $json);
     }
 
+    public function provideCannotAttachUserWithExistingRole()
+    {
+        return [
+            'same role' => ['publicationAdmins', 'publicationAdmins'],
+            'different role' => ['publicationAdmins', 'editors'],
+        ];
+    }
+
     /**
+     * @dataProvider provideCannotAttachUserWithExistingRole
      * @return void
      */
-    public function testUserRoleAndUserMustBeUniqueForAPublication()
+    public function testCannotAttachUserWithExistingRole($existingRole, $newRole)
     {
         $user = User::factory()->create();
-        $role_id = Role::where('name', Role::PUBLICATION_ADMINISTRATOR)->first()->id;
 
-        $publication = Publication::factory()->hasAttached(
-            $user,
-            [
-                'role_id' => $role_id,
-            ]
-        )
+        /** @var \App\Models\Publication $publication */
+        $publication = Publication::factory()
+            ->hasAttached($user, [], $existingRole)
             ->create();
-        $this->expectException(QueryException::class);
-        $publication->users()->attach(
-            $user,
-            [
-                'role_id' => $role_id,
-            ]
-        );
-        $publication_pivot_data = PublicationUser::where(
-            [
-                'user_id' => $user->id,
-                'role_id' => $role_id,
-                'publication_id' => $publication->id,
-            ]
-        )
-            ->get();
-        $this->assertEquals(1, $publication_pivot_data->count());
-    }
 
-    /**
-     * @return array
-     */
-    public function createPublicationUserViaMutationAsAnEditorProvider(): array
-    {
-        //@codingStandardsIgnoreStart
-        return [
-            //User Role ID,                     Allowed?
-            [ Role::SUBMITTER_ROLE_ID,          false ],
-            [ Role::REVIEWER_ROLE_ID,           false ],
-            [ Role::REVIEW_COORDINATOR_ROLE_ID, false ],
-            [ Role::EDITOR_ROLE_ID,             false ],
-            [ 0,                                false ], //TODO: These should be tested in the context of validating role ids, not repeatedly as a possible input
-            [ '',                               false ],
-            [ null,                             false ],
-        ];
-        //@codingStandardsIgnoreEnd
+        $this->expectException(QueryException::class);
+        $publication->$newRole()->attach($user);
```

Comment:
Should there also be an assertion that the role assignment worked as expected?

Reply:
This entire test is checking that the role assignments /didn't/ work, hence expectException.  Later tests assert the case when an attachment is successful.

Reply:
I forgot that `expectException()` is performed prior to the operation. This makes sense now.